### ISSUE TITLE
Nimbus Besu connection disruption

### ISSUE THREAD

**Issue by yorickdowne** [hyperledger/besu#5281]
### Description

Companion issue to https://github.com/status-im/nimbus-eth2/issues/4773, as it's unclear to me whether it's a Nimbus issue, a Besu issue, or both

Nimbus 23.3.2 with Besu 23.1.2, the Engine API connection gets disrupted. Reported on Discord; I think I can reproduce this. Fresh Nimbus/Besu sync on Goerli, with Nimbus checkpoint synced and Besu snap synced.

### Additional Information (Add any of the following or anything else that may be relevant)

Debug logs are too large to attach, placed them on dropbox: https://www.dropbox.com/s/69sejeefa9i85kt/nimbus-besu-connection-woes.debug.log.tar.gz?dl=0

Log examples from Nimbus
```
consensus_1             | WRN 2023-03-29 20:25:07.252+00:00 Connection to EL node degraded             topics="elmon" url=http://execution:8551 failedRequest=getBlockByNumber statusCode=0 err="{\"code\":-32603,\"message\":\"Internal error\"}"
consensus_1             | INF 2023-03-29 20:25:09.810+00:00 Connection to EL node restored             topics="elmon" url=http://execution:8551
```

and from Besu
```
execution_1             | java.net.SocketException: Connection reset
execution_1             |     at java.base/sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:394)
execution_1             |     at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:426)
execution_1             |     at io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:258)
execution_1             |     at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1132)
execution_1             |     at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:357)
execution_1             |     at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:151)
execution_1             |     at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
execution_1             |     at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
execution_1             |     at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
execution_1             |     at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
execution_1             |     at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
execution_1             |     at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
execution_1             |     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
execution_1             |     at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by non-fungible-nelson** [comment]
Do you have any notes on reproduction? Or frequency?

**Comment by yorickdowne** [comment]
I reproduced it by fresh-syncing a Nimbus:Besu pair on Goerli. Showed within a few hours.

I'll try a second pair overnight on a better machine, in case this is related to low performance of the host.

**Comment by stefa2k** [comment]
I've searched 14 days of logs of Besu and don't see this issue on Besu with Nimbus.

Do you have some more specifics, like server specs, configs?

**Comment by yorickdowne** [comment]
After some more testing, this may be related to test system 1 being a Potato (Netcup VPS). Test system two, a baremetal, has not shown the same errors.

Which then begs the question of whether that's worth your while to dig into. Please let me know.

**Comment by siladu** [comment]
@yorickdowne I can see from the logs that besu was still syncing during this period (both block import and snapsync world state download). That may have an impact on besu's ability to respond to RPCs. I will check the expected behaviour of eth_getByBlockNumber during a sync...it shouldn't be throwing an error though :)

We are missing logging of this exact error, so am adding that: https://github.com/hyperledger/besu/pull/5288

Have you got any metrics from this period, e.g. the "Besu Full" grafana dashboard?

---

Analysis of the logs...

```
execution_1             | 2023-03-29 20:22:06.387+00:00 | nioEventLoopGroup-3-6 | DEBUG | AbstractHandshakeHandler | Handshake error:
execution_1             | java.net.SocketException: Connection reset
```

^ This error is related to peer handshake not CL, so think we can ignore.


```
execution_1             | 2023-03-29 20:51:38.012+00:00 | EthScheduler-Services-34 (importBlock) | DEBUG | TransactionLogBloomCacher | Caching logs bloom for block 0x4c1892.
```

^ I see this log surrounding all of the calls to `eth_getBlockByNumber` and there are 443746 entries in the logs, the final one being 0x4c2704 (4990724). 
I'm pretty sure this much cache population only happens during startup of besu, so maybe it temporarily increases the resource requirements on besu and is another reason why something went south on a lower spec machine.

```
cat nimbus-besu-connection-woes.debug.log/nimbus-besu-connection-woes.nimbus.debug.log | grep  "getBlockByNumber" | grep "Internal error" | wc -l
      20
```

^ Number of errors seen on nimbus side.

```
cat nimbus-besu-connection-woes.debug.log/nimbus-besu-connection-woes.besu.debug.log | grep "eth_getBlockByNumber" | awk '{print $15$16}' | sort | uniq -c
  26 eth_getBlockByNumber["0x42A3DA",false]
 135 eth_getBlockByNumber["latest",false]
```

^ These are the requests we see on besu side. It doesn't tally exactly (maybe logs aren't exactly the same timeframe?) but matches more closely with `eth_getBlockByNumber["0x42A3DA",false]` which is a request this block https://goerli.beaconcha.in/block/4367322
which is from `759 days 17 hrs ago (Mar-01-2021)`

I'm not sure why nimbus would need to repeatedly ask for this block...maybe deposit contract related? I thought goerli deposit contract was much later than that though.

Also, if I hit this block on a fully sync goerli node, it returns a result without an issue.
I am yet to test what happens if you ask for it during syncing at the same point where your node was, but early on in the sync, it just returns a valid response containing a null value (since the syncing chain hasn't imported that block yet).

**Comment by siladu** [comment]
> Netcup VPS

@yorickdowne have you got more details about the spec please - I would like to try to recreate a something similar in AWS

**Comment by yorickdowne** [comment]
Hey @siladu thanks for digging! Yes I think you're right that low-spec machine + still syncing + a lot of requests caused the issues. Now that it's synced, it's not showing errors.

So on some level - this is acceptable. Potato may show errors while syncing. Fine. If it's something you can optimize fairly easily great, but optimization can be a bottomless hole to throw time into.

This is a Netcup VPS 4000 G9. 8 virtual cores on shared CPU - it does not get the compute from an actual 8 cores. 32 GiB memory. 800GB storage, shared access. It'll take me an hour to test IOPS.

VPS4000 G10, the current model, is 12 virtual cores, 24 GiB memory, same storage.

In AWS, maybe a t3a.xlarge with gp3 8k IOPS, and I'll adjust the IOPS number when I have it, see above.

**Comment by stefa2k** [comment]
8 vCPUs are almost always too short in my experience, especially with AWS

**Comment by siladu** [comment]
I span up a fleet of nodes (one per CL) with the extra logging from #5288 and may have recreated it with prysm and lighthouse on sepolia. Unfortunately the nimbus sepolia and nimbus goerli nodes didn't sync yet (no simple checkpoint sync flag, think it's a manual process) so will have to fix them up on Monday.

Leaving the error I found here to continue analysing next week...
```
java.lang.NullPointerException: Cannot invoke \"org.hyperledger.besu.ethereum.core.BlockHeader.getHash()\" because \"headHeader\" is null
at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.EthGetBlockByNumber.latestResult(EthGetBlockByNumber.java:86)
at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.AbstractBlockParameterMethod.findResultByParamType(AbstractBlockParameterMethod.java:90)
at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.AbstractBlockParameterMethod.response(AbstractBlockParameterMethod.java:102)
at org.hyperledger.besu.ethereum.api.jsonrpc.execution.BaseJsonRpcProcessor.process(BaseJsonRpcProcessor.java:42)
at org.hyperledger.besu.ethereum.api.jsonrpc.execution.TracedJsonRpcProcessor.process(TracedJsonRpcProcessor.java:41)
at org.hyperledger.besu.ethereum.api.jsonrpc.execution.TimedJsonRpcProcessor.process(TimedJsonRpcProcessor.java:45)
at org.hyperledger.besu.ethereum.api.jsonrpc.execution.AuthenticatedJsonRpcProcessor.process(AuthenticatedJsonRpcProcessor.java:51)
at org.hyperledger.besu.ethereum.api.jsonrpc.execution.JsonRpcExecutor.execute(JsonRpcExecutor.java:92)
at org.hyperledger.besu.ethereum.api.handlers.JsonRpcExecutorHandler.executeRequest(JsonRpcExecutorHandler.java:118)
at org.hyperledger.besu.ethereum.api.handlers.JsonRpcExecutorHandler.executeJsonObjectRequest(JsonRpcExecutorHandler.java:131)
at org.hyperledger.besu.ethereum.api.handlers.JsonRpcExecutorHandler.lambda$handler$0(JsonRpcExecutorHandler.java:78)
at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
at java.base/java.lang.Thread.run(Thread.java:833)\n"}
```

It might be as simple as not having the requested block number in the chain...although my earlier tests showed this should return a null field in an otherwise valid response, so something else is going on I think.

**Comment by siladu** [comment]
> In AWS, maybe a t3a.xlarge with gp3 8k IOPS, and I'll adjust the IOPS number when I have it, see above.

Thanks. Will see if I can recreate the nimbus scenario with something lower specced like this next week, including the new logging.
I used `m5.xlarge` (wasn't trying to low ball it).

**Comment by siladu** [comment]
@yorickdowne This was indeed a besu bug, related to a mismatch of cached vs non-cached data, essentially a race condition exposed by syncing (a rapid succession of block imports) and I guess lower-powered machines (slower to import individual blocks).

Thanks for the quality report!

Fixed by #5292

**Comment by yorickdowne** [comment]
EBS is kinda a potato, even at 16k IOPS. Even more so below that.
